internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,159 @@
+	flushingMemBuffer *MemDB
+	len, size         int // len and size records the total flushed and onflushing memdb.
+}
+
+const (
+	MinFlushKeys = 10000
+	MinFlushSize = 16 * 1024 * 1024  // 16MB
+	MaxFlushSize = 128 * 1024 * 1024 // 128MB
+)
+
+type FlushFunc func(*MemDB) error
+
+func NewPipelinedMemDB(flushFunc FlushFunc) *PipelinedMemDB {
+	return &PipelinedMemDB{
+		memBuffer: newMemDB(),
+		errCh:     make(chan error, 1),
+		flushFunc: flushFunc,
+	}
+}
+
+// Dirty returns whether the pipelined buffer is mutated.
+func (p *PipelinedMemDB) Dirty() bool {
+	return p.memBuffer.Dirty() || p.flushingMemBuffer != nil
+}
+
+// Get the value by given key, it returns tikverr.ErrNotExist if not exist.
+// The priority of the value is MemBuffer > flushingMemBuffer MemBuffer > Snapshot.
+func (p *PipelinedMemDB) Get(k []byte) ([]byte, error) {
+	v, err := p.memBuffer.Get(k)
+	if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
+		v, err = p.flushingMemBuffer.Get(k)
+	}
+	if err != nil {
+		return v, err
+	}
+	if len(v) == 0 {
+		return nil, tikverr.ErrNotExist
+	}
+	return v, nil
+}
+
+func (p *PipelinedMemDB) GetFlags(k []byte) (kv.KeyFlags, error) {
+	f, err := p.memBuffer.GetFlags(k)
+	if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
+		f, err = p.flushingMemBuffer.GetFlags(k)
+	}
+	if err != nil {
+		return 0, err
+	}
+	return f, nil
+}
+
+// MayFlush is called during execution of a transaction, it does flush when there are enough keys and the ongoing flushingMemBuffer is done.
+func (p *PipelinedMemDB) MayFlush() error {
+	if p.flushFunc == nil {
+		return nil
+	}
+	size := p.memBuffer.Size()
+	if size < MinFlushSize || (p.memBuffer.Len() < MinFlushKeys && size < MaxFlushSize) {
+		return nil
+	}
+	if !p.onFlushing.CompareAndSwap(false, true) && size < MaxFlushSize {
+		return nil
+	}
+	if p.flushingMemBuffer != nil {
+		if err := <-p.errCh; err != nil {
+			return err
+		}
+		if size >= MaxFlushSize {
+			// if the flush is triggered by size too large, the CAS operation may fail, manually set onFlushing to true in this case.
+			p.onFlushing.Store(true)
+		}
+	}
+	p.flushingMemBuffer = p.memBuffer
+	p.len += p.flushingMemBuffer.Len()
+	p.size += p.flushingMemBuffer.Size()
+	p.memBuffer = newMemDB()
+	go func() {
+		metrics.TiKVPipelinedFlushLenHistogram.Observe(float64(p.flushingMemBuffer.Len()))
+		metrics.TiKVPipelinedFlushSizeHistogram.Observe(float64(p.flushingMemBuffer.Size()))
+		err := p.flushFunc(p.flushingMemBuffer)
+		p.onFlushing.Store(false)
+		// Send the error to errCh after onFlushing status is set to false.
+		// this guarantees the onFlushing.Store(true) in another goroutine's MayFlush happens after onFlushing.Store(false) in this function.
+		p.errCh <- err
+	}()
+	return nil
+}
+
+// Flush will force flushingMemBuffer all the mutations, if there is an ongoing flush memdb, it will wait.
+// this should be called before running commit protocol.
+func (p *PipelinedMemDB) Flush() error {
```

Comment:
If we view the memdb in a layer under the transaction client, the function had better not assume anything on how it is used, like "this should be called before running commit protocol", at least the interface(name and comments) does not prevent upper users from using it elsewhere. The current implementation skips stuff like making a new memdb because of this assumption.
Maybe we could implement a function like doFlush(), and let MayFlush and ForceFlush both call it.

Reply:
The comment "this should be called before running commit protocol" is for reminding client-go developer.

Upper users can use it when they want to flush all the mutations into stores and wait for it.

Maybe the interface can be:

```
MayFlush() error // may flush in another goroutine, returns error if there are any failed flushing.
ForceFlush() error // force flush when it's not empty in another goroutine, returns error if there are any failed flushing.
Wait() error // wait the flush result if there is onflushing memdb.
```

Reply:
I changed the interface to:

```go
// It always flush in another goroutine.
// Flush(true) should be invoked in the end of a transaction.
// Upper layer users can use Flush(false) or Flush(true) as they want.
Flush(force bool) (bool, error)
FlushWait() error
```

This leaves the optimize change for multi-statement usage:

* Call `Flush(true)` in the end of a DML
* Call `FlushWait()` before the next statement's execution
* Then we can ensure all the mutations are flushed into stores and skip union scan.